### ISSUE TITLE
[QUESTION]Memory increase issue introduced by Megatron-FSDP during training.

### ISSUE THREAD

**Issue by chinapanda** [NVIDIA/Megatron-LM#4069]
I used the same Qwen3 script to train with Megatron-FSDP on the core_v0.14.0 and core_v0.16.0 branches of Megatron-LM, and found that the core_v0.16.0 branch consumes one-third more memory than core_v0.14.0. Can you provide any explanation for this?

qwen3脚本
torchrun \
    --nproc_per_node 8 \
    --nnodes 1 \
    --node_rank 0 \
    --master_addr localhost \
    --master_port 29600 pretrain_gpt.py \
    --data-path /data/qwen-data/openweb-qwen_text_document \
    --tokenizer-model /data/models/Qwen2-57B-A14B \
    --num-layers 48 \
    --hidden-size 2048 \
    --num-attention-heads 32 \
    --ffn-hidden-size 6144 \
    --num-query-groups 4 \
    --group-query-attention \
    --attention-softmax-in-fp32 \
    --distributed-backend nccl \
    --attention-dropout 0 \
    --hidden-dropout 0 \
    --use-rotary-position-embeddings \
    --rotary-base 1000000 \
    --num-experts 128 \
    --moe-router-topk 8 \
    --kv-channels 128 \
    --moe-token-dispatcher-type alltoall \
    --moe-ffn-hidden-size 768 \
    --moe-router-load-balancing-type aux_loss \
    --moe-aux-loss-coeff 0.001 \
    --untie-embeddings-and-output-weights \
    --swiglu \
    --normalization RMSNorm \
    --qk-layernorm \
    --disable-bias-linear \
    --lr 5e-6 \
    --lr-decay-style cosine \
    --min-lr 5e-7 \
    --adam-beta1 0.9 \
    --adam-beta2 0.95 \
    --adam-eps 1e-5 \
    --weight-decay 0.1 \
    --clip-grad 1 \
    --lr-warmup-iters 0 \
    --tokenizer-type HuggingFaceTokenizer \
    --split 949,50,1 \
    --data-cache-path ./cache \
    --log-throughput \
    --log-interval 1 \
    --save-interval 10000 \
    --eval-interval 1000 \
    --no-load-optim \
    --no-load-rng \
    --make-vocab-size-divisible-by 100 \
    --position-embedding-type rope \
    --no-position-embedding \
    --use-distributed-optimizer \
    --use-flash-attn \
    --norm-epsilon 1e-6 \
    --no-save-optim \
    --lr-decay-iters 1000 \
    --moe-grouped-gemm \
    --bf16 \
    --train-iters 10 \
    --eval-iters 1 \
    --tensor-model-parallel-size 1 \
    --pipeline-model-parallel-size 1 \
    --context-parallel-size 1 \
    --expert-model-parallel-size 1 \
    --seq-length 4096 \
    --max-position-embeddings 40960 \
    --micro-batch-size 1 \
    --global-batch-size 32 \
    --transformer-impl transformer_engine \
    --moe-router-pre-softmax \
    --no-gradient-accumulation-fusion \
    --use-megatron-fsdp  \
    --data-parallel-sharding-strategy optim_grads  \
    --ckpt-format fsdp_dtensor  \
    --use-distributed-optimizer \
    --num-layers 6 \
    --manual-gc \
    --manual-gc-interval 300

**Comment by janEbert** [comment]
@NVIDIA/megatron-fsdp maybe you have an idea what could have led to this?

**Comment by shjwudp** [comment]
Hi @chinapanda , I haven’t observed whether the `optim_grads` sharding strategy causes any noticeable memory growth. However, there’s a known bug in TransformerEngine versions v2.10 and v2.11 where FusedAdam saves the optimizer states as full-sized tensors instead of sharded DTensors as intended. Could you please check whether you also changed the TransformerEngine version during your test?

**Comment by chinapanda** [comment]
Yes, I upgraded to  TE2.11. Compared to not using Megatron-FSDP, using Megatron-FSDP results in more memory allocation in FusedAdam. The additional memory used by optim_grads and optim_grads_param primarily comes from excessive allocation of temporary space.

Below is the memory allocation under the four modes, which clearly reveals this issue.

Not using Megatron-FSDP.

<img width="1790" height="458" alt="Image" src="https://github.com/user-attachments/assets/0475039b-4626-4339-a1d6-66a8b04fde31" />

Optim
<img width="1805" height="878" alt="Image" src="https://github.com/user-attachments/assets/27eac829-23fb-482a-a068-c571a31423a9" />

Optim_grads
<img width="1712" height="503" alt="Image" src="https://github.com/user-attachments/assets/9e9c09e8-704b-43f6-841a-1826b45ebb07" />

Optim_grads_params
<img width="1706" height="623" alt="Image" src="https://github.com/user-attachments/assets/469a4e1d-6b23-476c-91a2-f1cb90d5cf92" />

**Comment by shjwudp** [comment]
I see. Could you please upgrade to TE v2.12? That version fixes the issue caused by the bugged FusedAdam implementation.

**Comment by ericharper** [comment]
@cspades can you give instructions to reproduce the fix?

**Comment by cspades** [comment]
@chinapanda I think I noticed this issue too - specifically that FusedAdam would initialize master weights when Megatron-FSDP already has its own copy of master weights, I _**think**_ this was fixed by this PR: https://github.com/NVIDIA/Megatron-LM/pull/4133 (but I made many many other fixes in MCore, so I would use ToT Megatron-LM to pick up at least this most recent fix related to FusedAdam: https://github.com/NVIDIA/Megatron-LM/pull/4802)

The fix is in `optimizers.__init__.py`:
```
                # NOTE(@cspades): FusedAdam is provided Megatron-FSDP's main weights as
                # non-quantized DTensor(s). Megatron-FSDP should NEVER use FusedAdam's
                # main weights, complete waste of memory as the optimizer step is still
                # applied to the Megatron-FSDP main weight and extended to FusedAdam
                # main weights. Override this here.
                setattr(optimizer_part.optimizer, "master_weights", False)
```

As for TE, all of my fixes have gone in 2.14 or 2.15, there was an optimizer state fix related to FusedAdam losing its DTensors.

Can you try ToT Megatron-Core? If you need help with the version bump feel free to ask, this memory usage was a really dumb bug.

**Comment by cspades** [comment]
@chinapanda Did my suggestion above work? It's the closest idea I had from a quick look at your Torch memory profiles and that extra master weight is created during state initialization on the first step.

Will close this issue if it's been resolved!
